patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs

```diff
@@ -202,217 +192,93 @@ internal partial class UIModBrowser : UIState, IHaveBackButtonCommand
-					using (var reader = new StreamReader(zip))
-						modlist = JArray.Parse(reader.ReadToEnd());
-				}
-				else {
-					// Fallback if needed.
-					modlist = (JArray)jsonObject["modlist"];
-				}
-				foreach (var mod in modlist.Children<JObject>()) _items.Add(UIModDownloadItem.FromJson(installedMods, mod));
-				UpdateNeeded = true;
-			}
-			catch (Exception e) {
-				LogModBrowserException(e);
-			}
+		internal void InnerPopulateModBrowser() {
+			// Populate
+			SteamWorkshop = new WorkshopHelper.QueryHelper();
+			Items.AddRange(SteamWorkshop.QueryWorkshop());
+			UpdateNeeded = true;
 		}
 
 		/// <summary>
 		///     Enqueues a list of mods, if found on the browser (also used for ModPacks)
 		/// </summary>
 		internal void DownloadMods(IEnumerable<string> modNames) {
-			var downloads = new List<DownloadFile>();
+			var downloads = new List<UIModDownloadItem>();
 
 			foreach (string desiredMod in modNames) {
-				var mod = _items.FirstOrDefault(x => x.ModName == desiredMod);
+				var mod = Items.FirstOrDefault(x => x.ModName == desiredMod);
 				if (mod == null) { // Not found on the browser
 					_missingMods.Add(desiredMod);
 				}
 				else if (mod.Installed == null || mod.HasUpdate) { // Found, add to downloads
-					var modDownload = mod.GetModDownload();
-					downloads.Add(modDownload);
+					downloads.Add(mod);
 				}
 			}
 
 			// If no download detected for some reason (e.g. empty modpack filter), prevent switching UI
 			if (downloads.Count <= 0) return;
 
-			SoundEngine.PlaySound(SoundID.MenuTick);
-			Interface.downloadProgress.gotoMenu = Interface.modBrowserID;
-			Interface.downloadProgress.OnDownloadsComplete += () => {
-				if (_missingMods.Count > 0) {
-					Interface.infoMessage.Show(Language.GetTextValue("tModLoader.MBModsNotFoundOnline", string.Join(",", _missingMods)), Interface.modBrowserID);
-				}
-				_missingMods.Clear();
-			};
+			WorkshopHelper.ModManager.Download(downloads);
 
-			Interface.downloadProgress.HandleDownloads(downloads.ToArray());
+			if (_missingMods.Count > 0) {
+				Interface.infoMessage.Show(Language.GetTextValue("tModLoader.MBModsNotFoundOnline", string.Join(",", _missingMods)), Interface.modBrowserID);
+			}
+			_missingMods.Clear();
 		}
 
-		internal UIModDownloadItem FindModDownloadItem(string modName)
-			=> _items.FirstOrDefault(x => x.ModName.Equals(modName));
-
-		//		private void OnModDownloadCompleted(HttpDownloadRequest req) {
-		//			if (req.Response.StatusCode != HttpStatusCode.OK) {
-		//				string errorKey = req.Response.StatusCode == HttpStatusCode.ServiceUnavailable ? "MBExceededBandwidth" : "MBUnknownMBError";
-		//				Interface.errorMessage.Show(Language.GetTextValue("tModLoader." + errorKey), 0);
-		//			}
-		//			else if (req.Success && req.CustomData is UIModDownloadItem currentDownload) {
-		//				ProcessDownloadedMod(req, currentDownload);
-		//			}
-		//		}
+		internal UIModDownloadItem FindModDownloadItem(string publishId)
```

Comment:
Revert this, and instead just box PublishId dependency within Workshop implement, not browser.
Mods should be based on traditional ModReferences by modname